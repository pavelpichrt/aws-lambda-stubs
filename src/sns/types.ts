import { SNSEventRecord, SNSMessageAttributes } from 'aws-lambda';

export interface SNSEventParams {
  Message?: string;
  TopicArn?: string;
  Subject?: string;
  EventSubscriptionArn?: string;
  MessageAttributes?: SNSMessageAttributes;
  MessageId?: string;
  Timestamp?: string;
}

export interface SNSEventGenerator {
  ({
    Subject,
    Message,
    TopicArn,
    EventSubscriptionArn,
    MessageAttributes,
    MessageId,
    Timestamp,
  }?: SNSEventParams): SNSEventRecord;
}
