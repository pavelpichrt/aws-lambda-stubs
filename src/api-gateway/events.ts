import { v4 as uuidv4 } from 'uuid';

import { generateRandomInt, randomAlphaNumericString } from '../support';
import {
  APIGatewayAuthorizerEventGenerator,
  APIGatewayEventContextGenerator,
  APIGatewayEventGenerator,
} from './types';

const generateEventContext: APIGatewayEventContextGenerator = ({
  resourceId,
  authorizer,
  resourcePath,
  path,
  stage,
  apiId,
  userAgent,
  sourceIp,
  httpMethod = 'POST',
}) => {
  const actualAuthorizer = authorizer || {
    principalId: 'mock-principal',
    integrationLatency: generateRandomInt(5, 500),
  };
  const monthNames = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  const now = new Date();

  return {
    resourceId: resourceId || randomAlphaNumericString(6),
    authorizer: actualAuthorizer,
    resourcePath,
    httpMethod,
    extendedRequestId: `${randomAlphaNumericString(15)}=`,
    requestTime: `${now.getDate()}/${
      monthNames[now.getMonth()]
    }/${now.getFullYear()}:${now.getHours()}:${now.getMinutes()}:${now.getSeconds()} +0000`,
    path,
    accountId: '123456278901',
    protocol: 'HTTP/1.1',
    stage,
    domainPrefix: apiId,
    requestTimeEpoch: now.getTime(),
    requestId: uuidv4(),
    identity: {
      cognitoIdentityPoolId: null,
      apiKey: null,
      apiKeyId: null,
      clientCert: null,
      accountId: null,
      cognitoIdentityId: null,
      caller: null,
      sourceIp,
      principalOrgId: null,
      accessKey: null,
      cognitoAuthenticationType: null,
      cognitoAuthenticationProvider: null,
      userArn: null,
      userAgent,
      user: null,
    },
    domainName: `${apiId}.execute-api.eu-west-2.amazonaws.com`,
    apiId,
  };
};

export const apiGatewayProxyEvent: APIGatewayEventGenerator = ({
  resourceId,
  authorizer,
  stage = 'mock-stage',
  resourcePath = '/mock/{id}/path',
  apiId = 'mockapi',
  userAgent = 'MockApi/1.0.0',
  sourceIp = '123.123.123.123',
  pathParameters,
  queryStringParameters = null,
  body,
  headers = {
    'Content-Type': 'application/json',
  },
  authorizationHeaderGenerator,
  includeAuthorizationHeader = true,
  path = `/${stage}/${resourcePath}`,
  httpMethod = 'POST',
} = {}) => {
  const actualPathParameters: Record<string, string> = {};
  (resourcePath.match(/({[a-zA-Z0-9_-]+})/g) || []).forEach(
    (fullParam: string) => {
      const paramWithoutBrackets = fullParam.substring(1, fullParam.length - 1);
      const paramValue = pathParameters?.[paramWithoutBrackets] || 'foo';

      path = path.replace(fullParam, paramValue);
      actualPathParameters[paramWithoutBrackets] = paramValue;
    },
  );
  const additionalHeaders: Record<string, string> = headers;

  if (includeAuthorizationHeader) {
    const defaultHeaderGenerator = () => randomAlphaNumericString(32);
    const actualAuthorizationHeaderGenerator =
      authorizationHeaderGenerator || defaultHeaderGenerator;

    additionalHeaders.Authorization = actualAuthorizationHeaderGenerator();
  }

  const eventContext = generateEventContext({
    resourceId,
    authorizer,
    resourcePath,
    path,
    stage,
    apiId,
    userAgent,
    sourceIp,
    httpMethod,
  });
  const traceId = `Root=1-${randomAlphaNumericString(
    8,
  )}-${randomAlphaNumericString(24)}`;
  const additionalHeadersMultiValue: Record<string, string[]> = {};

  for (const [headerName, headerValue] of Object.entries(additionalHeaders)) {
    additionalHeadersMultiValue[headerName] = [headerValue];
  }

  return {
    resource: resourcePath,
    path,
    httpMethod,
    queryStringParameters,
    headers: {
      Accept: '*/*',
      'Accept-Encoding': 'gzip, deflate, br',
      Host: `${apiId}.execute-api.eu-west-2.amazonaws.com`,
      'User-Agent': userAgent,
      'X-Amzn-Trace-Id': traceId,
      'X-Forwarded-For': sourceIp,
      'X-Forwarded-Port': '443',
      'X-Forwarded-Proto': 'https',
      ...additionalHeaders,
    },
    multiValueHeaders: {
      Accept: ['*/*'],
      'Accept-Encoding': ['gzip, deflate, br'],
      Host: [`${apiId}.execute-api.eu-west-2.amazonaws.com`],
      'User-Agent': [userAgent],
      'X-Amzn-Trace-Id': [traceId],
      'X-Forwarded-For': [sourceIp],
      'X-Forwarded-Port': ['443'],
      'X-Forwarded-Proto': ['https'],
      ...additionalHeadersMultiValue,
    },
    multiValueQueryStringParameters: null,
    pathParameters: actualPathParameters,
    stageVariables: null,
    requestContext: eventContext,
    body: body ? JSON.stringify(body) : null,
    isBase64Encoded: false,
  };
};

export const apiGatewayAuthorizerEvent: APIGatewayAuthorizerEventGenerator = ({
  authorizationToken = 'authorization-token-stub',
  resource = '/prod/POST/{proxy+}',
} = {}) => ({
  type: 'TOKEN',
  authorizationToken,
  methodArn: `arn:aws:execute-api:eu-west-1:123456789012:example${resource}`,
});
