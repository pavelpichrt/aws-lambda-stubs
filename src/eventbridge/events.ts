import { v4 as uuidv4 } from 'uuid';

import { EventBridgeEventGenerator } from './types';

export const eventBridgeEvent: EventBridgeEventGenerator = ({
  version = '1',
  source = 'aws.events',
  accountId = '123456789012',
  time = new Date().toISOString(),
  region = 'eu-west-1',
  resources,
  detail = {},
} = {}) => ({
  id: uuidv4(),
  version,
  'detail-type': 'Scheduled Event',
  source,
  account: accountId,
  time,
  region,
  resources: resources || [
    `arn:aws:events:${region}:${accountId}:rule/ExampleRule`,
  ],
  detail,
});
