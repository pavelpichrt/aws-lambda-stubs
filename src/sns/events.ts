import { v4 as uuidv4 } from 'uuid';
import { SNSEventGenerator, SNSEventParams } from './types';

const generateSnsEventObject: SNSEventGenerator = ({
  Subject = 'EXAMPLE-SUBJECT',
  Message = 'EXAMPLE-MESSAGE',
  TopicArn = 'arn:aws:sns:eu-west-1-1:123456789012:ExampleTopic',
  EventSubscriptionArn = 'arn:aws:sns:us-east-1:123456789012:ExampleTopic',
  MessageAttributes = {
    Test: {
      Type: 'String',
      Value: 'TestString',
    },
    TestBinary: {
      Type: 'Binary',
      Value: 'TestBinary',
    },
  },
  MessageId = uuidv4(),
  Timestamp = new Date().toISOString(),
} = {}) => ({
  EventSource: 'aws:sns',
  EventVersion: '1.0',
  EventSubscriptionArn,
  Sns: {
    SignatureVersion: '1',
    Timestamp,
    Signature: 'EXAMPLE',
    SigningCertUrl: 'EXAMPLE',
    MessageId,
    Message,
    MessageAttributes,
    Type: 'Notification',
    UnsubscribeUrl: 'EXAMPLE-UNSUBSCRIBE-URL',
    TopicArn,
    Subject,
  },
});

export const snsEvent = (parametrizedEvents: SNSEventParams[]) => ({
  Records: parametrizedEvents.map(generateSnsEventObject),
});
