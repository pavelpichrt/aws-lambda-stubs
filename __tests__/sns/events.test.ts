import { snsEvent } from '../../src/sns/events';
import { uuidRegex, isoDateRegex } from '../config';

describe('snsEvent', () => {
  it('Generates a correct stub', () => {
    const {
      Records: [
        {
          Sns: { MessageId, Timestamp, ...snsDeterministicProps },
          ...deterministicProps
        },
      ],
    } = snsEvent([{}]);

    expect(deterministicProps).toMatchSnapshot();
    expect(snsDeterministicProps).toMatchSnapshot();
    expect(MessageId).toEqual(expect.stringMatching(uuidRegex));
    expect(Timestamp).toEqual(expect.stringMatching(isoDateRegex));
  });
});
