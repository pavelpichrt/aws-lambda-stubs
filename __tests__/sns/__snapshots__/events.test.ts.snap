// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snsEvent Generates a correct stub 1`] = `
{
  "EventSource": "aws:sns",
  "EventSubscriptionArn": "arn:aws:sns:us-east-1:123456789012:ExampleTopic",
  "EventVersion": "1.0",
}
`;

exports[`snsEvent Generates a correct stub 2`] = `
{
  "Message": "EXAMPLE-MESSAGE",
  "MessageAttributes": {
    "Test": {
      "Type": "String",
      "Value": "TestString",
    },
    "TestBinary": {
      "Type": "Binary",
      "Value": "TestBinary",
    },
  },
  "Signature": "EXAMPLE",
  "SignatureVersion": "1",
  "SigningCertUrl": "EXAMPLE",
  "Subject": "EXAMPLE-SUBJECT",
  "TopicArn": "arn:aws:sns:eu-west-1-1:123456789012:ExampleTopic",
  "Type": "Notification",
  "UnsubscribeUrl": "EXAMPLE-UNSUBSCRIBE-URL",
}
`;
